### PR TITLE
Include "got" and "expected" details in method mismatch errors

### PR DESCRIPTION
**Pull request by jchavarri**
This PR adds more details to the end of the error message, detailing the difference between the expected and actual type of the method that failed. The code follows the same approach than the `Errortrace.Escape` branch.

### CODE REVIEW

File: typing/printtyp.ml
Comment (gasche) [inline typing/printtyp.ml]: The message here is now very similiar to the other `Incompatible_fields` case above (currently on lines 2451-2457.) Could you share code between the two with a new `explain_incompatible_fields` helper function? (This means producing the same message in both cases, which is also nice for consistency. Personally I think that the message above is slightly better as it says the same thing with less words.)

**Review by gasche — COMMENTED** [review]
The result seems nice to me (but an okay from @Octachron would be helpful). See a minor comment inline about the implementation.

File: typing/printtyp.ml
Comment (jchavarri) [reply typing/printtyp.ml]: Done in https://github.com/ocaml/ocaml/pull/12777/commits/b01dfaeb432d5b63bf19109ae66fa363e7903cdc.

**Review by gasche — APPROVED** [review]
I am happy with the implementation and believe this is an improvement, approved. Before merging, I would still prefer to wait for a green light of @Octachron who knows the overall picture of error message -- it is unlikely, but there might be a reason why those two cases were handled separately.

**Comment by gasche** [comment]
(Please include a Changes entry: this is a small change but a nice contribution, error messages need all the help they can get.)

File: testsuite/tests/typing-modules/inclusion_errors.ml
Comment (Octachron) [inline testsuite/tests/typing-modules/inclusion_errors.ml:900]: The variable names are incoherent with the previous line which is a sign that there was not enough naming context sharing.

Comment (gasche) [reply testsuite/tests/typing-modules/inclusion_errors.ml:900]: It took me several reads to decipher what you wrote, and I would guess that it will be obscure to @jchavarri as well. I think that you are pointing out that there is a bug in the implementation, related to the handling of the printing context. This results in `'a . ... as 'a` above being printed as `'b . ... as 'b` below, when the naming choice should remain consistent.

Comment (gasche) [reply testsuite/tests/typing-modules/inclusion_errors.ml:900]: (If this is correct, I would be interested in your help to fix it, because it is very unclear for someone not familiar with these internal details. I looked and my guess is that there is no obvious fix because the code needs to be massaged a bit to do things correctly.)

Comment (jchavarri) [reply testsuite/tests/typing-modules/inclusion_errors.ml:900]: Thanks for the explanations. I will take a stab at fixing the problem, and come up with hopefully specific questions if I find any blockers.

Comment (Octachron) [reply testsuite/tests/typing-modules/inclusion_errors.ml:900]: @gasche Sorry for being unclear, I am looking at it; I just wanted to point out that there was some issues with the complicated invariants of printtyp before you merge the PR.

File: typing/printtyp.ml
Comment (Octachron) [inline typing/printtyp.ml:2446]: The `type_expr_with_reserved_names` function should be called on type expressions whose explicit type variable names have been reserved.

This is not the case in the new branch, I propose to move the
```ocaml
  reserve_names diff.got;
  reserve_names diff.expected;
```
part in this function to avoid this.

File: testsuite/tests/typing-modules/inclusion_errors.ml
Comment (Octachron) [reply testsuite/tests/typing-modules/inclusion_errors.ml:900]: A full fix seems more involved than I thought, I propose to just fix the precondition on `type_expr_with_reserved_names` for now.

File: typing/printtyp.ml
Comment (jchavarri) [reply typing/printtyp.ml:2446]: Makes sense, done in https://github.com/ocaml/ocaml/pull/12777/commits/392a30c3c235860d2c893759ead6ef7d37d867ad.

Just curious, and I assume this is out of the scope, but would it make sense to move the call to `reserve_names` inside `type_expr_with_reserved_names` itself?

Comment (Octachron) [reply typing/printtyp.ml:2446]: This would be suboptimal: `reserve_names` (and `prepare_types`) compute metadata to decide how to print loops in type expressions and how to disambiguate type variable names. 

In both case, it is better to have the full picture available before starting to print.

Otherwise we might decide for instance to name `'b` an anonymous type variable in the first printed type expression whereas in the second the user was using `'b` explicitly and we end-up needed to rename the user-chosen variable names to account for our greedy choice of anonymous type variables.

**Review by Octachron — APPROVED** [review]

**Comment by Octachron** [comment]
Merged, thanks for the improvement to error messages !

**Pull request merged**
